### PR TITLE
DPE-839: Introducing spark-shell wrapper. Tests and scala.shell.histfile config merging coming soon.

### PR DESCRIPTION
**Pull request by averma-canonical**
We want to create a wrapper for the “spark-shell” command, building on the parsing of configurations and merging logic already implemented for spark-submit

Main features:
* autodetect k8s master
* setting the .scala_history file location dynamically (refinement in override behavior in progress)
    $SNAP_USER_DATA/.scala_history if SNAP_USER_DATA defined OR
    $HOME/.scala_history

UPDATE:
Added logic to merge special java options like configurations which was broken, revealed for .scala_history needed for spark-shell

### CODE REVIEW

File: helpers/spark-shell.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+USER_HOME_DIR_ENT_IDX = 5
+
+if __name__ == "__main__":
+    logging.basicConfig(
+        format="%(asctime)s %(levelname)s %(message)s", level=logging.INFO
+    )
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--username", default=None, type=str, help="Username of service account to use."
```
Comment (averma-canonical) [inline helpers/spark-shell.py:19]: NOTE: Regarding parity with spark-submit, following optional parameters have not been added.

--deploy-mode - Discussed with Enrico. Not appropriate here.
--log-level - Will piggyback with other review comments changes

File: tests/unittest/test_helpers.py [hunk 2]
```diff
@@ -8,5 +12,162 @@ def test_dummy(self):
         pass
 
 
+class TestProperties(UnittestWithTmpFolder):
```
Comment (deusebio) [inline tests/unittest/test_helpers.py:15]: That's very good we already have tests

Comment (averma-canonical) [reply tests/unittest/test_helpers.py:15]: Thanks Enrico! Felt this was necessary especially after I got burnt during the demo!

**Review by welpaolo — DISMISSED** [review]
Looks good to me. I left some minor comments mainly related to the documentation.

You can also add a doc string to briefly describe the purpose of each test.

File: helpers/utils.py [hunk 3]
```diff
@@ -28,8 +28,30 @@
 PathLike = Union[str, "os.PathLike[str]"]
 
 
+def is_property_with_options(key: str) -> bool:
+    """Check if a given property is known to be options-like requiring special parsing.
+
+    Args:
+        key: Property for which special options-like parsing decision has to be taken
+    """
+    return key in ["spark.driver.extraJavaOptions"]
```
[the remark this answers is not in the document]
Comment (deusebio) [reply helpers/utils.py:37]: +1

File: helpers/utils.py [hunk 4]
```diff
@@ -28,8 +28,30 @@
 PathLike = Union[str, "os.PathLike[str]"]
 
 
+def is_property_with_options(key: str) -> bool:
+    """Check if a given property is known to be options-like requiring special parsing.
```
[the remark this answers is not in the document]
Comment (deusebio) [reply helpers/utils.py:32]: No @welpaolo . Documentation should be imperative! Please have a look at this PEP
 
https://peps.python.org/pep-0257/#one-line-docstrings

File: helpers/utils.py [hunk 5]
```diff
@@ -28,8 +28,30 @@
...
+
+    Args:
+        key: Property for which special options-like parsing decision has to be taken
+    """
+    return key in ["spark.driver.extraJavaOptions"]
+
+
+def get_properties_with_options(conf: Dict) -> Dict:
+    """Extract properties which are known to be options-like requiring special parsing.
```
[the remark this answers is not in the document]
Comment (deusebio) [reply helpers/utils.py:41]: as before. Keep the imperative. I have strong views on this, given the choice of imperative form is supported by PEP

File: helpers/utils.py [hunk 6]
```diff
@@ -76,14 +102,58 @@ def print_properties(props: Dict) -> None:
         print(f"{k}={v}")
 
 
-def merge_configurations(dictionaries_to_merge: List[Dict]) -> Dict:
+def parse_options(options_string: Optional[str]) -> Dict:
```
[the remark this answers is not in the document]
Comment (deusebio) [reply helpers/utils.py:105]: +1

File: tests/integration/ie-tests.sh [hunk 7]
```diff
@@ -1,14 +1,17 @@
...
+}
 
+test_example_job() {
   spark-client.setup-spark-k8s service-account
 
   KUBE_CONFIG=/home/${USER}/.kube/config
 
   K8S_MASTER_URL=k8s://$(kubectl --kubeconfig=${KUBE_CONFIG} config view -o jsonpath="{.clusters[0]['cluster.server']}")
-  SPARK_EXAMPLES_JAR_NAME='spark-examples_2.12-3.4.0-SNAPSHOT.jar'
+  #SPARK_EXAMPLES_JAR_NAME='spark-examples_2.12-3.4.0-SNAPSHOT.jar'
```
Comment (deusebio) [inline tests/integration/ie-tests.sh:13]: Remove comments

**Review by deusebio — APPROVED** [review]
Only some minors and support for (some) @welpaolo comments

File: helpers/utils.py [hunk 8]
```diff
@@ -54,7 +74,7 @@ def read_property_file(name: Optional[str]) -> Dict:
 
 
 def write_property_file(fp: io.TextIOWrapper, props: Dict, log: bool = None) -> None:
-    """Write a given dictionary to provided file descriptor.
+    """Writes a given dictionary to provided file descriptor.
```
Comment (deusebio) [inline helpers/utils.py:77]: Imperative

File: helpers/utils.py [hunk 9]
```diff
@@ -76,14 +96,81 @@ def print_properties(props: Dict) -> None:
...
+def merge_dictionaries(dictionaries_to_merge: List[Dict]) -> Dict:
     """Merges a given list of dictionaries, properties in subsequent ones override the properties in prior ones in the dictionary list."""
     result = dict()
     for override in dictionaries_to_merge:
         result.update(override)
     return result
 
 
+def merge_options(dictionaries_to_merge: List[Dict]) -> Dict:
+    """Merges a given list of options, options in subsequent ones override the options in prior ones in the list.
```
Comment (deusebio) [inline helpers/utils.py:146]: Imperative

**Review by deusebio — APPROVED** [review]
Good to go!

File: helpers/utils.py [hunk 10]
```diff
@@ -11,25 +11,41 @@
...
+
+    Args:
+        key: Property for which special options-like parsing decision has to be taken
+    """
+    return key in [
+        constants.OPTION_SPARK_DRIVER_DEFAULT_JAVA_OPTIONS,
+        constants.OPTION_SPARK_DRIVER_EXTRA_JAVA_OPTIONS,
+        constants.OPTION_SPARK_EXECUTOR_DEFAULT_JAVA_OPTIONS,
+        constants.OPTION_SPARK_EXECUTOR_EXTRA_JAVA_OPTIONS,
+    ]
```
Comment (welpaolo) [inline helpers/utils.py:31]: you can even create a constant that contains this list.

**Review by welpaolo — APPROVED** [review]

File: helpers/utils.py (hunk 10 above)
Comment (averma-canonical) [reply helpers/utils.py:31]: Sure, will address this separately

**Pull request merged**